stats/opentelemetry/trace.go

```diff
@@ -0,0 +1,125 @@
+	countRecvMsg        uint32
+	previousRpcAttempts uint32
+}
+
+// traceTagRPC populates context with a new span, and serializes information
+// about this span into gRPC Metadata.
+func (csh *clientStatsHandler) traceTagRPC(ctx context.Context, rti *stats.RPCTagInfo) (context.Context, *attemptTraceSpan) {
+	if csh.options.TraceOptions.TextMapPropagator == nil {
+		return ctx, nil
+	}
+
+	mn := "Attempt." + strings.Replace(removeLeadingSlash(rti.FullMethodName), "/", ".", -1)
+	tracer := otel.Tracer("grpc-open-telemetry")
+	ctx, span := tracer.Start(ctx, mn)
+
+	carrier := otelinternaltracing.NewCustomCarrier(ctx) // Use internal custom carrier to inject
+	otel.GetTextMapPropagator().Inject(ctx, carrier)
+
+	return carrier.Context(), &attemptTraceSpan{
+		span:         span,
+		countSentMsg: 0, // msg events scoped to scope of context, per attempt client side
+		countRecvMsg: 0,
+	}
+}
+
+// traceTagRPC populates context with new span data, with a parent based on the
+// spanContext deserialized from context passed in (wire data in gRPC metadata)
+// if present.
+func (ssh *serverStatsHandler) traceTagRPC(ctx context.Context, rti *stats.RPCTagInfo) (context.Context, *attemptTraceSpan) {
+	if ssh.options.TraceOptions.TextMapPropagator == nil {
+		return ctx, nil
+	}
+
+	mn := strings.Replace(removeLeadingSlash(rti.FullMethodName), "/", ".", -1)
+	var span trace.Span
+	tracer := otel.Tracer("grpc-open-telemetry")
+	ctx = otel.GetTextMapPropagator().Extract(ctx, otelinternaltracing.NewCustomCarrier(ctx))
+	// If the context.Context provided in `ctx` to tracer.Start(), contains a
+	// Span then the newly-created Span will be a child of that span,
+	// otherwise it will be a root span.
+	ctx, span = tracer.Start(ctx, mn, trace.WithSpanKind(trace.SpanKindServer))
+	return ctx, &attemptTraceSpan{
+		span:         span,
+		countSentMsg: 0,
+		countRecvMsg: 0,
+	}
+}
+
+// populateSpan populates span information based on stats passed in, representing
+// invariants of the RPC lifecycle. It ends the span, triggering its export.
+// This function handles attempt spans on the client-side and call spans on the
+// server-side.
+func populateSpan(_ context.Context, rs stats.RPCStats, ti *attemptTraceSpan) {
+	if ti == nil || ti.span == nil {
+		// Shouldn't happen, tagRPC call comes before this function gets called
+		// which populates this information.
+		logger.Error("ctx passed into stats handler tracing event handling has no span present")
+		return
+	}
+	span := ti.span
+
+	switch rs := rs.(type) {
+	case *stats.Begin:
+		// Note: Go always added Client and FailFast attributes even though they are not
+		// defined by the OpenCensus gRPC spec. Thus, they are unimportant for
+		// correctness.
+		span.SetAttributes(
+			attribute.Bool("Client", rs.Client),
+			attribute.Bool("FailFast", rs.Client),
+			attribute.Int64("previous-rpc-attempts", int64(ti.previousRpcAttempts)),
+			attribute.Bool("transparent-retry", rs.IsTransparentRetryAttempt),
```

Comment:
how is IsTransparentRetry being populated from rpc?

Reply:
Already replied here
https://github.com/aranjans/grpc-go/pull/4#discussion_r1816047989